lighthouse-core/audits/uses-rel-preconnect.js

```diff
@@ -0,0 +1,150 @@
+        `connections to important third-party origins. [Learn more](https://developers.google.com/web/fundamentals/performance/resource-prioritization#preconnect).`,
+      requiredArtifacts: ['devtoolsLogs'],
+      scoreDisplayMode: Audit.SCORING_MODES.NUMERIC,
+    };
+  }
+
+  /**
+   * Check if record has valid timing
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {!boolean}
+   */
+  static hasValidTiming(record) {
+    return record.timing && record.timing.connectEnd > 0 && record.timing.connectStart > 0;
+  }
+
+  /**
+   * Check is the connection is already open
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {!boolean}
+   */
+  static hasAlreadyConnectedToOrigin(record) {
+    return (
+      record.timing.dnsEnd - record.timing.dnsStart === 0 &&
+      record.timing.connectEnd - record.timing.connectStart === 0
+    );
+  }
+
+  /**
+   * Check is the connection has started before the socket idle time
+   * @param {!WebInspector.NetworkRequest} record
+   * @param {!WebInspector.NetworkRequest} mainResource
+   * @return {!boolean}
+   */
+  static socketStartTimeIsBelowThreshold(record, mainResource) {
+    return Math.max(0, record._startTime - mainResource._endTime) < PRECONNECT_SOCKET_MAX_IDLE;
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static async audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[UsesRelPreconnectAudit.DEFAULT_PASS];
+    let maxWasted = 0;
+
+    const [networkRecords, mainResource] = await Promise.all([
+      artifacts.requestNetworkRecords(devtoolsLogs),
+      artifacts.requestMainResource(devtoolsLogs),
+    ]);
+
+    const origins = networkRecords
+      .filter(record => {
+        return (
+          // filter out all resources where timing info was invalid
+          UsesRelPreconnectAudit.hasValidTiming(record) &&
+          // filter out all resources that are loaded by the document
+          record.initiatorRequest() !== mainResource &&
+          // filter out all resources that have the same origin
+          mainResource.parsedURL.securityOrigin() !== record.parsedURL.securityOrigin() &&
+          // filter out urls that do not have an origin (data, ...)
+          !!record.parsedURL.securityOrigin() &&
+          // filter out all resources where origins are already resolved
+          !UsesRelPreconnectAudit.hasAlreadyConnectedToOrigin(record) &&
+          // make sure the requests are below the PRECONNECT_SOCKET_MAX_IDLE (15s) mark
+          UsesRelPreconnectAudit.socketStartTimeIsBelowThreshold(record, mainResource)
+        );
+      })
+      .map(record => record.parsedURL.securityOrigin());
+
+    const preconnectOrigins = new Set(origins);
+    let results = [];
+    preconnectOrigins.forEach(origin => {
+      const records = networkRecords.filter(record => record.parsedURL.securityOrigin() === origin);
```

Comment:
we need all records that survived the above filtering. 

So one way we could do it is to use a `Map` for this `origins` variable above. The key is the securityOrigin, and the value is an an array of networkRecords (that survived the filter).